From e9e259bccbd5781309dd78f36ab8ad0b44305dbe Mon Sep 17 00:00:00 2001
From: Zach Brown <zab@redhat.com>
Date: Fri, 17 Oct 2014 22:06:41 -0400
Subject: [fs] buffer: increase the buffer-head per-CPU LRU size

Message-id: <1413583601-11165-3-git-send-email-zab@redhat.com>
Patchwork-id: 97662
O-Subject: [RHEL7.1 PATCH 2/2] fs/buffer.c: increase the buffer-head per-CPU LRU size
Bugzilla: 1053108
RH-Acked-by: Ian Kent <ikent@redhat.com>
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>
RH-Acked-by: Rik van Riel <riel@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1053108
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8115844

Signed-off-by: Zach Brown <zab@redhat.com>

Backport of the following upstream commit:

commit 86cf78d73de8c6bfa89804b91ee0ace71a459961
Author: Sebastien Buisson <sebastien.buisson@bull.net>
Date:   Thu Oct 9 15:29:38 2014 -0700

    fs/buffer.c: increase the buffer-head per-CPU LRU size

    Increase the buffer-head per-CPU LRU size to allow efficient filesystem
    operations that access many blocks for each transaction.  For example,
    creating a file in a large ext4 directory with quota enabled will access
    multiple buffer heads and will overflow the LRU at the default 8-block LRU
    size:

    * parent directory inode table block (ctime, nlinks for subdirs)
    * new inode bitmap
    * inode table block
    * 2 quota blocks
    * directory leaf block (not reused, but pollutes one cache entry)
    * 2 levels htree blocks (only one is reused, other pollutes cache)
    * 2 levels indirect/index blocks (only one is reused)

    The buffer-head per-CPU LRU size is raised to 16, as it shows in metadata
    performance benchmarks up to 10% gain for create, 4% for lookup and 7% for
    destroy.

    Signed-off-by: Liang Zhen <liang.zhen@intel.com>
    Signed-off-by: Andreas Dilger <andreas.dilger@intel.com>
    Signed-off-by: Sebastien Buisson <sebastien.buisson@bull.net>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/buffer.c b/fs/buffer.c
index 25de151..2be98bc 100644
--- a/fs/buffer.c
+++ b/fs/buffer.c
@@ -1248,7 +1248,7 @@ static struct buffer_head *__bread_slow(struct buffer_head *bh)
  * a local interrupt disable for that.
  */
 
-#define BH_LRU_SIZE 8
+#define BH_LRU_SIZE 16
 
 struct bh_lru {
  struct buffer_head *bhs[BH_LRU_SIZE];
-- 
1.7.1